src/SFML/Graphics/LineShape.cpp

```diff
@@ -0,0 +1,95 @@
+////////////////////////////////////////////////////////////
+//
+// SFML - Simple and Fast Multimedia Library
+// Copyright (C) 2007-2022 Laurent Gomila (laurent@sfml-dev.org)
+//
+// This software is provided 'as-is', without any express or implied warranty.
+// In no event will the authors be held liable for any damages arising from the use of this software.
+//
+// Permission is granted to anyone to use this software for any purpose,
+// including commercial applications, and to alter it and redistribute it freely,
+// subject to the following restrictions:
+//
+// 1. The origin of this software must not be misrepresented;
+//    you must not claim that you wrote the original software.
+//    If you use this software in a product, an acknowledgment
+//    in the product documentation would be appreciated but is not required.
+//
+// 2. Altered source versions must be plainly marked as such,
+//    and must not be misrepresented as being the original software.
+//
+// 3. This notice may not be removed or altered from any source distribution.
+//
+////////////////////////////////////////////////////////////
+
+////////////////////////////////////////////////////////////
+// Headers
+////////////////////////////////////////////////////////////
+#include <SFML/Graphics/LineShape.hpp>
+
+
+namespace sf
+{
+////////////////////////////////////////////////////////////
+LineShape::LineShape(const Vector2f& point1, const Vector2f& point2, float thickness)
+    : m_direction(point2 - point1)
+    , m_thickness(thickness)
+{
+    setPosition(point1);
+    update();
+}
+
+
+////////////////////////////////////////////////////////////
+void LineShape::setThickness(float thickness)
+{
+    m_thickness = thickness;
+    update();
+}
+
+
+////////////////////////////////////////////////////////////
+void LineShape::setEndPoint(const Vector2f& point)
+{
+    m_direction = point - getPosition();
+    update();
+}
+
+
+////////////////////////////////////////////////////////////
+float LineShape::getThickness() const { return m_thickness; }
+
+
+////////////////////////////////////////////////////////////
+Vector2f LineShape::getEndPoint() const { return getPosition() + m_direction; }
+
+
+////////////////////////////////////////////////////////////
+float LineShape::getLength() const { return m_direction.length(); }
+
+
+////////////////////////////////////////////////////////////
+size_t LineShape::getPointCount() const { return 4; }
+
+
+////////////////////////////////////////////////////////////
+Vector2f LineShape::getPoint(size_t index) const
+{
+    auto offset = Vector2f();
+    if (m_direction != Vector2f(0, 0))
+        offset = (m_thickness / 2.f) * m_direction.normalized().perpendicular();
```

Comment:
Instead of caching each point (as discussed in the other comment), it's also possible to just cache `offset`. 
Then, we:
* don't have to check for zero vector
* can do a simple vector addition in `getPoint()`

Reply:
I noticed that other shapes don't do any caching. Is this something we want to start doing with all shapes?

If we do cache offset, the only time it would get invalidated is when `setThickness` or `setEndPoint` are called, right?

Reply:
I'm working through adding caching of the offset but I don't see how we get around checking `m_direction` for zero-length. At some point the offset has to be calculated and it'll change depending on whether `m_direction` has length.

Reply:
I'm also fine with not caching things, and only restart this discussion once we have evidence that performance is indeed an issue. Since these shapes tend to be used for debug graphics, it's well possible that an uncached implementation is completely fine. And it's less complex (regarding invalidation, etc), meaning less error-prone.

> I'm working through adding caching of the offset but I don't see how we get around checking m_direction for zero-length.

True, what I meant is that we don't have to check it each frame, only when direction changes.

Reply:
> I'm also fine with not caching things, and only restart this discussion once we have evidence that performance is indeed an issue. 

This was my thought as well. If other shapes can get away without caching, then I think it's safe to avoid caching in `LineShape`.